File: src/includes/odrmanagerlite.jl
export
    OdrManagerLite,
    Position,

    loadfile,
    printdata,
    create_position,
    activate_position,

    get_trackpos,
    get_lanepos,
    get_inertialpos,
    get_footpoint,

    set_pos,
    set_trackpos,
    set_lanepos,
    set_inertialpos,

    track2inertial,
    inertial2track,
    lane2inertial,
    inertial2lane,
    footpoint2inertial,

    print_odrmanagerlite,
    get_curvature,
    get_track_len,
    get_lane_width

type OdrManagerLite
    ptr::Ptr{Void}
    has_activated_position::Bool
    
    function OdrManagerLite()
        ptr = ccall( (:createOdrManagerLite, LIB_ODRMGR), Ptr{Void}, () )
        odrmanager = new(ptr, false)
        finalizer(odrmanager, obj -> begin
            ccall( (:free_OdrManagerLite, LIB_ODRMGR), Void, (Ptr{Void},), obj.ptr )
        end)
        odrmanager
    end
end

type Position
    ptr::Ptr{Void}
    Position(ptr::Ptr{Void}) = new(ptr)
end

loadfile(mgr::OdrManagerLite, name::String) =
    ccall((:odr_manager_loadFile, LIB_ODRMGR), Bool, (Ptr{Void}, Ptr{UInt8}), mgr.ptr, name)

printdata(mgr::OdrManagerLite) =
    ccall((:odr_manager_printData, LIB_ODRMGR), Void, (Ptr{Void},), mgr.ptr )

function create_position(mgr::OdrManagerLite)
    mgr.has_activated_position = true
    ptr = ccall((:odr_manager_createPosition, LIB_ODRMGR), Ptr{Void}, (Ptr{Void},), mgr.ptr)
    Position(ptr)
end

function activate_position(mgr::OdrManagerLite, pos::Position)
    mgr.has_activated_position = true
    ccall((:odr_manager_activatePosition, LIB_ODRMGR), Void, (Ptr{Void}, Ptr{Position}), mgr.ptr, pos.ptr)
end

function get_trackpos(mgr::OdrManagerLite)
    if !mgr.has_activated_position
        warn("OdrManagerLite does not have an activated position")
    else
        coord = TrackCoord()
        ccall((:odr_manager_getTrackPos, LIB_ODRMGR), Void, (Ptr{Void},Ptr{Void}), mgr.ptr, pointer_from_objref(coord))
        return coord
    end
end
function get_lanepos(mgr::OdrManagerLite)
    if !mgr.has_activated_position
        warn("OdrManagerLite does not have an activated position")
    else
        coord = LaneCoord()
        ccall((:odr_manager_getLanePos, LIB_ODRMGR), Void, (Ptr{Void},Ptr{Void}), mgr.ptr, pointer_from_objref(coord))
        return coord
    end
end
function get_inertialpos(mgr::OdrManagerLite)
    if !mgr.has_activated_position
        warn("OdrManagerLite does not have an activated position")
    else
        coord = Coord()
        ccall((:odr_manager_getInertialPos, LIB_ODRMGR), Void, (Ptr{Void},Ptr{Void}), mgr.ptr, pointer_from_objref(coord))
        return coord
    end
end
function get_footpoint(mgr::OdrManagerLite)
    if !mgr.has_activated_position
        warn("OdrManagerLite does not have an activated position")
    else
        coord = Coord()
        ccall((:odr_manager_getFootPoint, LIB_ODRMGR), Void, (Ptr{Void},Ptr{Void}), mgr.ptr, pointer_from_objref(coord))
        return coord
    end
end

set_pos(mgr::OdrManagerLite, value::TrackCoord) =
    ccall((:odr_manager_setpos_track_coord, LIB_ODRMGR), Void, (Ptr{Void}, Ptr{TrackCoord}), mgr.ptr, pointer_from_objref(value))
set_pos(mgr::OdrManagerLite, value::LaneCoord) =
    ccall((:odr_manager_setpos_lane_coord, LIB_ODRMGR), Void, (Ptr{Void}, Ptr{LaneCoord}), mgr.ptr, pointer_from_objref(value))
set_pos(mgr::OdrManagerLite, value::Coord) =
    ccall((:odr_manager_setpos_coord, LIB_ODRMGR), Void, (Ptr{Void}, Ptr{Coord}), mgr.ptr, pointer_from_objref(value))

set_trackpos(mgr::OdrManagerLite, value::TrackCoord) =
    ccall((:odr_manager_set_track_pos_track_coord, LIB_ODRMGR), Void, (Ptr{Void}, Ptr{TrackCoord}), mgr.ptr, pointer_from_objref(value))
set_trackpos(mgr::OdrManagerLite, id::Integer, s::Real, t::Real=0.0) =
    ccall((:odr_manager_set_track_pos_s_t, LIB_ODRMGR), Void, (Ptr{Void}, Cint, Cdouble, Cdouble), mgr.ptr, id, s, t)

set_lanepos(mgr::OdrManagerLite, value::LaneCoord) =
    ccall((:odr_manager_setLanePos_with_lanecoord, LIB_ODRMGR), Void, (Ptr{Void}, Ptr{LaneCoord}), mgr.ptr, pointer_from_objref(value))
set_lanepos(mgr::OdrManagerLite, trackId::Integer, laneId::Integer, s::Real, offset::Real=0.0) =
    ccall((:odr_manager_setLanePos, LIB_ODRMGR), Void, (Ptr{Void}, Cint, Cint, Cdouble, Cdouble), mgr.ptr, trackId, laneId, s, offset)

set_inertialpos(mgr::OdrManagerLite, x::Real, y::Real, z::Real) =
    ccall((:odr_manager_setInertialPos, LIB_ODRMGR), Void, (Ptr{Void}, Cdouble, Cdouble, Cdouble), mgr.ptr, x, y, z)

track2inertial(mgr::OdrManagerLite) = 
     ccall((:odr_manager_track2inertial, LIB_ODRMGR), Bool, (Ptr{Void},), mgr.ptr)

inertial2track(mgr::OdrManagerLite) = 
     ccall((:odr_manager_inertial2track, LIB_ODRMGR), Bool, (Ptr{Void},), mgr.ptr)

lane2inertial(mgr::OdrManagerLite) = 
     ccall((:odr_manager_lane2inertial, LIB_ODRMGR), Bool, (Ptr{Void},), mgr.ptr)

inertial2lane(mgr::OdrManagerLite) = 
     ccall((:odr_manager_inertial2lane, LIB_ODRMGR), Bool, (Ptr{Void},), mgr.ptr)

footpoint2inertial(mgr::OdrManagerLite) = 
    ccall((:odr_manager_footPoint2inertial, LIB_ODRMGR), Void, (Ptr{Void},), mgr.ptr)

print_odrmanagerlite(mgr::OdrManagerLite, indent::Integer=0) = 
    ccall((:odr_manager_print, LIB_ODRMGR), Void, (Ptr{Void},Cint), mgr.ptr, indent)

get_curvature(mgr::OdrManagerLite) = 
     ccall((:odr_manager_getCurvature, LIB_ODRMGR), Cdouble, (Ptr{Void},), mgr.ptr)
    
get_track_len(mgr::OdrManagerLite, trackId::Integer) =
    ccall((:odr_manager_getTrackLen, LIB_ODRMGR), Cdouble, (Ptr{Void}, Cint), mgr.ptr, trackId)

get_lane_width(mgr::OdrManagerLite) =
    ccall((:odr_manager_getLaneWidth, LIB_ODRMGR), Cdouble, (Ptr{Void},), mgr.ptr)

File: deps/compile/odrmanagerlite.cpp
void * createOdrManagerLite()
{
    OpenDrive::OdrManagerLite * mgr = new OpenDrive::OdrManagerLite();
    void * retval = mgr;
    return retval;
}
void free_OdrManagerLite(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    delete mgr;
}

void free_Position(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    delete mgr;
}

bool odr_manager_loadFile(void * void_odr_manager, char * odr_name)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    std::string name(odr_name);
    return mgr->loadFile(name);
}

void odr_manager_printData(void * void_odr_manager) 
{     
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    mgr->printData();
}
void * odr_manager_createPosition(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    return mgr->createPosition();
}
void odr_manager_activatePosition(void * void_odr_manager, void * void_pos)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::Position * pos = reinterpret_cast<OpenDrive::Position*>(void_pos);
    mgr->activatePosition(pos);
}
void odr_manager_getTrackPos(void * void_odr_manager, void * void_track_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::TrackCoord * coord = reinterpret_cast<OpenDrive::TrackCoord*>(void_track_coord);
    (*coord) = mgr->getTrackPos();
}
void odr_manager_getLanePos(void * void_odr_manager, void * void_lane_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::LaneCoord * coord = reinterpret_cast<OpenDrive::LaneCoord*>(void_lane_coord);
    (*coord) = mgr->getLanePos();
}
void odr_manager_getInertialPos(void * void_odr_manager, void * void_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::Coord * coord = reinterpret_cast<OpenDrive::Coord*>(void_coord);
    (*coord) = mgr->getInertialPos();
}
void odr_manager_getFootPoint(void * void_odr_manager, void * void_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::Coord * coord = reinterpret_cast<OpenDrive::Coord*>(void_coord);
    (*coord) = mgr->getFootPoint();
}
void odr_manager_setpos_track_coord(void * void_odr_manager, void * void_track_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::TrackCoord * coord = reinterpret_cast<OpenDrive::TrackCoord*>(void_track_coord);
    mgr->setPos(*coord);
}
void odr_manager_set_track_pos_s_t(void * void_odr_manager, int id, double s, double t)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    mgr->setTrackPos(id, s, t);
}
void odr_manager_set_track_pos_track_coord(void * void_odr_manager, void * void_track_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::TrackCoord * value = reinterpret_cast<OpenDrive::TrackCoord*>(void_track_coord);
    mgr->setTrackPos(*value);
}
void odr_manager_setpos_lane_coord(void * void_odr_manager, void * void_lane_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::LaneCoord * value = reinterpret_cast<OpenDrive::LaneCoord*>(void_lane_coord);
    mgr->setPos(*value);
}
void odr_manager_setLanePos(void * void_odr_manager, int trackId, int laneId, double s, double offset)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    mgr->setLanePos(trackId, laneId, s, offset);
}
void odr_manager_setLanePos_with_lanecoord(void * void_odr_manager, void * void_lane_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::LaneCoord * value = reinterpret_cast<OpenDrive::LaneCoord*>(void_lane_coord);
    mgr->setLanePos(*value);
}
void odr_manager_setpos_coord(void * void_odr_manager, void * void_coord)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    OpenDrive::Coord * value = reinterpret_cast<OpenDrive::Coord*>(void_coord);
    mgr->setPos(*value);
}
void odr_manager_setInertialPos(void * void_odr_manager, double x, double y, double z)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    mgr->setInertialPos(x, y, z);
}
bool odr_manager_track2inertial(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    return mgr->track2inertial();
}
bool odr_manager_inertial2track(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    return mgr->inertial2track();
}
bool odr_manager_lane2inertial(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    return mgr->lane2inertial();
}
bool odr_manager_inertial2lane(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    return mgr->inertial2lane();
}
void odr_manager_print(void * void_odr_manager, int ident)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    mgr->print(ident);
}
double odr_manager_getCurvature(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
     return mgr->getCurvature(); 
}
double odr_manager_getTrackLen(void * void_odr_manager, int trackId)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
     return mgr->getTrackLen(trackId); 
}
double odr_manager_getLaneWidth(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    return mgr->getLaneWidth();  
}
void odr_manager_footPoint2inertial(void * void_odr_manager)
{
    OpenDrive::OdrManagerLite * mgr = reinterpret_cast<OpenDrive::OdrManagerLite*>(void_odr_manager);
    mgr->footPoint2inertial();  
}

